src/main/scala/com/kinja/amqp/AmqpProducer.scala

```diff
@@ -16,24 +26,59 @@ import play.api.libs.json._
 
 import java.util.concurrent.TimeUnit
 
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.Try
+
 class AmqpProducer(
 	connection: ActorRef,
 	actorSystem: ActorSystem,
+	messageStore: MessageStore,
 	connectionTimeOut: Long,
+	askTimeout: Long,
 	logger: Slf4jLogger)(exchange: ExchangeParameters) {
 
 	private val channel: ActorRef = createChannel()
 
 	Amqp.waitForConnection(actorSystem, connection, channel).await(connectionTimeOut, TimeUnit.SECONDS)
 
-	def publish[A: Writes](routingKey: String, message: A): Unit = {
+	channel ! ConfirmSelect
+	channel ! AddConfirmListener(createConfirmListener)
+
+	def publish[A: Writes](routingKey: String, message: A, saveTimeMillis: Long = System.currentTimeMillis()): Future[Try[Unit]] = {
```

Comment:
`Try` also has `recover` and `recoverWith`. The ultimate approach here is to have two separate layers and switch between the transformer and non-transformer versions as you need to. `TryT[scalaz.Future, A]` is equivalent `scala.concurrent.Future` but it allows you to use `run` to have a finer degree of control over error-handling and also allows you to use `Try` in synchronous contexts and simply `liftM` when you need to combine with asynchronous code. This also allows you to have a separate type for failing and non-failing `Future`s.